### PR TITLE
New owned preallocated

### PR DESCRIPTION
**Pull request by dan-jfisher**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+unreleased
+  * Add `Buffer::new_owned_preallocated` function.
```
Comment (de-vri-es) [inline CHANGELOG.md]:
```suggestion
Unreleased:
  * Add `Buffer::new_owned_preallocated(...)` to transfer ownership of a pre-allocated buffer to Aravis.
```

File: aravis/src/manual/buffer.rs [hunk 2]
```diff
@@ -13,6 +14,29 @@ pub enum ImageError {
...
 impl Buffer {
+
+	/// Create an Aravis buffer that owns its own data from a pre-allocated raw buffer.
+	///
+	/// The created buffer's user data and destory callback are generated such that the provided
+	/// `destroy_callback` argument is called when the buffer is dropped.
+	///
+	/// # Safety
+	/// The resulting buffer borrows the data, but it carries no lifetime.
+	/// The user has to ensure the buffer stays valid.
```
Comment (de-vri-es) [inline aravis/src/manual/buffer.rs]:
```suggestion
	/// Create an Aravis buffer that owns its own data from a pre-allocated raw buffer.
	///
	/// The `destroy_callback` argument is called to destroy the buffer, and should free the resources associated with the buffer.
	///
	/// # Safety
	/// The pointer and length parameter must indicate a valid memory region where Aravis can safely write data to until the `destroy_callback` is called.
```

File: aravis/src/manual/buffer.rs [hunk 3]
```diff
@@ -23,8 +47,11 @@ impl Buffer {
 	/// The resulting buffer borrows the data, but it carries no lifetime.
 	/// The user has to ensure the buffer stays valid.
 	pub unsafe fn new_preallocated(data: *mut u8, len: usize) -> Self {
```
Comment (de-vri-es) [inline aravis/src/manual/buffer.rs:48]:
```suggestion
	pub unsafe fn new_preallocated(data: *mut u8, len: usize) -> Self {
```

Maybe we should rename this to `new_preallocated_borrowed()`, and keep `new_preallocated()` as a deprecated alias.

Also, if we do, I think `new_preallocated_borrowed()` and `new_preallocated_owned()` sounds better than `new_borrowed_preallocated()` and `new_owned_preallocated()`. Although I don't feel very strongly on this matter ^^

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+Unreleased:
+  * Add `Buffer::new_owned_preallocated(...)` to transfer ownership of a pre-allocated buffer to Aravis.
```
Comment (de-vri-es) [inline CHANGELOG.md]:
```suggestion
  * Add `Buffer::new_preallocated_owned(...)` to transfer ownership of a pre-allocated buffer to Aravis.
```

File: aravis/src/manual/buffer.rs [hunk 5]
```diff
@@ -13,6 +14,27 @@ pub enum ImageError {
...
 
 impl Buffer {
+
+	/// Create an Aravis buffer that owns its own data from a pre-allocated raw buffer.
+	///
+	/// The `destroy_callback` argument is called to destroy the buffer, and should free the resources associated with the buffer.
+	///
+	/// # Safety
+	/// The pointer and length parameter must indicate a valid memory region where Aravis can safely write data to until the `destroy_callback` is called.
+	pub fn new_preallocated_owned<F: FnOnce()>(data: *mut u8, len: usize, destroy_callback: F) -> Self {
```
Comment (de-vri-es) [inline aravis/src/manual/buffer.rs:24]: Shall we be nice and pass the data pointer to the callback?

Comment (dan-jfisher) [reply aravis/src/manual/buffer.rs:24]: Seems kind of confusing for the user that's writing the callback.  My thinking was that they should just move everything they need for the desctructor into a lambda and be done with it.

Maybe I am misunderstanding your suggestion though.

Comment (dan-jfisher) [reply aravis/src/manual/buffer.rs:24]: Are you suggesting that they write a callback such that `F: FnOnce(*mut u8)`?

Comment (de-vri-es) [reply aravis/src/manual/buffer.rs:24]: Yeah. You're right that it is hopefully normally not needed. In that sense I can see the logic in not providing it by default.

Fair enough, lets leave it out :)

Comment (dan-jfisher) [reply aravis/src/manual/buffer.rs:24]: I guess that does give the user slightly more flexibility.  Idk I'm conflicted.

File: aravis/src/manual/buffer.rs [hunk 6]
```diff
@@ -22,9 +44,26 @@ impl Buffer {
...
-		let buffer =
-			aravis_sys::arv_buffer_new_full(len, data as *mut c_void, std::ptr::null_mut(), None);
+		Self::preallocated(data as *mut c_void, len, std::ptr::null_mut(), None)
+	}
+
+	/// Create an Aravis buffer from a pre-allocated raw buffer.
+	///
+	/// The created buffer has no registered user data or destroy callback,
+	/// so management of the underlying buffer has to be done externally.
+	/// The buffer can be identified later when it is returned by a stream only byt the data pointer.
```
Comment (de-vri-es) [inline aravis/src/manual/buffer.rs]:
```suggestion
	/// The buffer can be identified later when it is returned by a stream only by the data pointer.
```

File: aravis/src/manual/buffer.rs [hunk 7]
```diff
@@ -22,9 +44,26 @@ impl Buffer {
...
+	///
+	/// # Safety
+	/// The resulting buffer borrows the data, but it carries no lifetime.
+	/// The user has to ensure the buffer stays valid.
```
Comment (de-vri-es) [inline aravis/src/manual/buffer.rs]:
```suggestion
	/// The pointer and length parameter must indicate a valid memory region where Aravis can safely write data to until the buffer is dropped.
	/// The created buffer borrows the memory buffer, but it carries no lifetime.
	/// See [`Self::new_preallocated_owned`] for a safer (but still unsafe) alternative.
```

File: aravis/src/manual/buffer.rs (hunk 5 above)
Comment (de-vri-es) [reply aravis/src/manual/buffer.rs:24]: It not really more flexible, just simpler if you actually need the data pointer. But you rarely need it.

If you really need it, you can capture a copy of it for the callback.

**Comment by de-vri-es** [comment]
Released as v0.8.1 :rocket:

**Pull request merged**